File: product-service/src/main/java/com/shopflow/product/service/ProductService.java
package com.shopflow.product.service;

import com.shopflow.product.dto.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface ProductService {

    List<CategoryResponse> listCategories();

    void updateAvgRating(java.util.UUID productId, java.math.BigDecimal avgRating);

    ProductImageResponse uploadProductImage(UUID sellerId, UUID productId,
                                            org.springframework.web.multipart.MultipartFile file, boolean primary);

    ProductResponse toggleProductStatus(UUID sellerId, UUID productId,
                                        boolean active, String role);

    ProductResponse createProduct(UUID sellerId, ProductRequest request);

    ProductResponse getProduct(UUID productId);

    ProductResponse updateProduct(UUID sellerId, UUID productId, ProductRequest request);

    // role passed as String to avoid coupling the service interface
    // to a security-layer enum. Caller extracts role from JWT and
    // passes it in. "SELLER" can only delete own products.
    // "ADMIN" can delete any product.
    void deleteProduct(UUID userId, UUID productId, String role);

    Page<ProductResponse> listProducts(int page, int size);

    Page<ProductSearchResponse> searchProducts(String query, int page, int size);

    Page<ProductResponse> getProductsBySeller(UUID sellerId, int page, int size);

    void updateStock(UUID skuId, int quantity, StockUpdateRequest.StockOperation operation);

    SkuDetailsResponse getSkuDetails(UUID skuId);

    Page<ProductResponse> getAllProducts(Pageable pageable);
    void deleteProductByAdmin(UUID productId);

}

File: shopflow-frontend/src/services/modules/product.service.js
import api from '../api/axiosInstance';

export const productService = {
  getProducts: async (page = 0, size = 20) => {
    const response = await api.get(`/products?page=${page}&size=${size}`);
    return response.data;
  },
  
  getProductById: async (id) => {
    const response = await api.get(`/products/${id}`);
    return response.data;
  },
  
  searchProducts: async (query, page = 0, size = 20) => {
    const response = await api.get(`/products/search?q=${query}&page=${page}&size=${size}`);
    return response.data;
  },
  
  getCategories: async () => {
    const response = await api.get('/categories');
    return response.data;
  },
  
  getProductsBySeller: async (sellerId, page = 0, size = 20) => {
    const response = await api.get(`/products/seller/${sellerId}?page=${page}&size=${size}`);
    return response.data;
  },
  
  // Seller endpoints
  createProduct: async (productData) => {
    return await api.post('/products', productData);
  },
  
  updateProduct: async (id, productData) => {
    return await api.put(`/products/${id}`, productData);
  },
  
  deleteProduct: async (id) => {
    return await api.delete(`/products/${id}`);
  },
  
  uploadImage: async (productId, file, primary = false) => {
    const formData = new FormData();
    formData.append('file', file);
    return await api.post(`/products/${productId}/images?primary=${primary}`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
  },
  
  toggleStatus: async (id, active) => {
    return await api.patch(`/products/${id}/status`, { active });
  }
};
